administrator/components/com_templates/Model/TemplateModel.php

```diff
@@ -148,6 +152,151 @@ public function getDirectoryTree($dir)
+		$filePath      = Path::clean($file);
+		$explodeArray  = explode(DIRECTORY_SEPARATOR, $filePath);
+
+		// Only allow html/ folder
+		if($explodeArray['1'] !== 'html')
+		{
+			$app->enqueueMessage(Text::_('COM_TEMPLATES_ERROR_ONLY_HTML_FOLDER_ALLOWED'), 'error');
+
+			return false;
+		}
+
+		$fileName = basename($filePath);
+		$type     = $explodeArray['2'];
+		$client   = ApplicationHelper::getClientInfo($client_id);
+
+		$componentPath = Path::clean($client->path . '/components/');
+		$modulePath    = Path::clean($client->path . '/modules/');
+		$layoutPath    = Path::clean(JPATH_ROOT . '/layouts/');
+
+		// For modules
+		if (stristr($type, 'mod_') != false)
+		{
+			$folder   = $explodeArray['2'];
+			$htmlPath = Path::clean($modulePath . $folder . '/tmpl/');
+			$fileName = $this->getSafeName($fileName);
+			$coreFile = $this->findPath($htmlPath, $fileName);
+
+			return $coreFile;
+		}
+		elseif (stristr($type, 'com_') != false)
+		{
+			// For components
+			$folder    = $explodeArray['2'];
+			$subFolder = $explodeArray['3'];
+			$fileName  = $this->getSafeName($fileName);
+
+			// The old scheme, if a view has a tmpl folder
+			$oldHtmlPath = Path::clean($componentPath . $folder . '/views/' . $subFolder . '/tmpl/');
+
+			if(!$coreFile = Path::find($oldHtmlPath, $fileName))
+			{
+				// The new scheme, the views are directly in the component/tmpl folder
+				$newHtmlPath = Path::clean($componentPath . $folder . '/tmpl/' . $subFolder . '/');
+				$coreFile    = $this->findPath($newHtmlPath, $fileName);
+
+				return $coreFile;
+			}
+
+			return $coreFile;
+		}
+		elseif (stristr($type, 'layouts') != false)
+		{
+			// For Jlayouts
+			$subtype = $explodeArray['3'];
+
+			if(stristr($subtype, 'com_'))
+			{
+				$folder    = $explodeArray['3'];
+		 		$subFolder = array_slice($explodeArray, 4, -1);
+				$subFolder = implode(DIRECTORY_SEPARATOR, $subFolder);
+				$htmlPath  = Path::clean($componentPath . $folder . '/layouts/' . $subFolder);
+				$fileName  = $this->getSafeName($fileName);
+				$coreFile  = $this->findPath($htmlPath, $fileName);
+
+				return $coreFile;
+			}
+			elseif (stristr($subtype, 'joomla') || stristr($subtype, 'libraries') || stristr($subtype, 'plugins'))
+			{
+				$subFolder = array_slice($explodeArray, 3, -1);
+				$subFolder = implode(DIRECTORY_SEPARATOR, $subFolder);
+				$htmlPath  = Path::clean($layoutPath . $subFolder);
+				$fileName  = $this->getSafeName($fileName);
+				$coreFile  = $this->findPath($htmlPath, $fileName);
+
+				return $coreFile;
+			}
+		}
+
+		$app->enqueueMessage(Text::_('COM_TEMPLATES_ERROR_CORE_FILE_NOT_FOUND'), 'error');
+
+		return false;
+	}
+
+	/**
+	 * Creates a safe file name for the given name.
+	 *
+	 * @param   string  $name  The filename
+	 *
+	 * @return  string $fileName  The filtered name without Date
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	private function getSafeName($name)
+	{
+		if (preg_match('/[0-9]/', $name))
```

Comment:
I like the fact that you make the diff view possible for the files with attached date.
Maybe we can do that even better.
At the moment, there must be a number in the name. You probably wanted to check if a date was attached. This test is in my opinion not necessary.
Already for the file /html/mod_search/default-red0-red.php the diff view modules/ mod_search/tmpl/default.php would be found.
Maybe we just have to look for a minus. Then a diff view would also be displayed for the files /html/mod_search/default-red.php and /html/mod_search/default-green.php - in case that the user renamed the file.

Reply:
![selection_016](https://user-images.githubusercontent.com/30978328/40405481-6cfe2b42-5e7a-11e8-9284-6ac3c157dd62.png)

@astridx Yeah, you are right my first approach is based on minus signs. But, it fails here because <code>action-button.php</code> is the core file.

Reply:
I think action button is the only file with a minus. All others are separated by an underscore. Is there a rule that a file should not contain a minus sign unless it is overwritten twice and the date is appended? Or can we ignore this one file (action-button)? So we could also offer the diff view for the doubly overwritten files. Otherwise, I do not see any rule that we can safely identify duplicate files. Then we should limit ourselves to the really equal written files.

Reply:
> Already for the file /html/mod_search/default-red0-red.php the diff view modules/ mod_search/tmpl/default.php would be found.

Is we have to treat <code>default-red0-red.php</code> as core file <code>default.php</code> ?

Reply:
![selection_017](https://user-images.githubusercontent.com/30978328/40408015-d05257dc-5e83-11e8-9a6e-d1d6f89454f9.png)

One more example.
@astridx In my function. I think we have to add one more check of date format . Then, it should fine :)

Reply:
May, this can help here http://php.net/manual/en/function.checkdate.php#113205